### PR TITLE
Updates to LM PR

### PR DESCRIPTION
**Pull request by priyakasimbeg**

### CODE REVIEW

File: algoperf/workloads/lm/lm_jax/workload.py [hunk 1]
```diff
@@ -101,41 +82,58 @@ def model_fn(
...
+    Returns:
+      {'summed': scalar summed loss, 'n_valid_examples': scalar number of
+      valid examples in batch, 'per_example': 1-d array of per-example losses}
+    """
+    if logits.ndim != targets.ndim + 1:
+      raise ValueError(
+        f'Incorrect shapes. Got shape {logits.shape} logits and '
+        f'{targets.shape} targets.'
+      )
+    smoothed_targets = optax.smooth_labels(
```
Comment (rka97) [inline algoperf/workloads/lm/lm_jax/workload.py]: Will calling onehot add a lot of extra memory? Since targets are already ids, we can instead use the old way without adding extra memory.

File: algoperf/workloads/lm/lm_jax/workload.py [hunk 2]
```diff
@@ -145,8 +143,10 @@ def _eval_batch(self,
     """Evaluate the model on a single batch."""
     logits, _ = self.model_fn(
         params, batch, model_state, spec.ForwardPassMode.EVAL, rng, False)
-    targets = batch['targets']
-
     # Calculate cross-entropy loss
-    loss = -jnp.sum(targets * jax.nn.log_softmax(logits, axis=-1))
-    return loss
+    # TODO(kasimbeg): add weights?
+    metrics = self.compute_weighted_cross_entropy(logits, batch['targets'], batch['weights'])
```
Comment (rka97) [inline algoperf/workloads/lm/lm_jax/workload.py]: For this workload I don't think we need weights for the cross-entropy calculation. Maybe we should explicitly del any weights?

File: algoperf/workloads/lm/input_pipeline.py [hunk 3]
```diff
@@ -1,154 +1,138 @@
...
+  ds = get_lm_dataset(data_rng, split, data_dir, global_batch_size, num_batches)
+  
+  it = map(
+    functools.partial(
+      data_utils.shard_and_maybe_pad_np, global_batch_size=global_batch_size
+    ),
+    ds,
+  )
+
+  return iter(it)
```
Comment (rka97) [inline algoperf/workloads/lm/input_pipeline.py:69]: Should we use itertools.cycle here?

File: algoperf/workloads/lm/lm_pytorch/workload.py
Comment (rka97) [inline algoperf/workloads/lm/lm_pytorch/workload.py:136]: We should remove _eval_batch in this file. The workload.py version already does the same thing.

File: algoperf/workloads/lm/lm_jax/workload.py [hunk 4]
```diff
@@ -101,41 +82,58 @@ def model_fn(
...
+  def _normalize_eval_metrics(
+    self, num_examples: int, total_metrics: Dict[str, Any]
+  ) -> Dict[str, float]:
+    """Normalize eval metrics."""
+    del num_examples
+    eval_denominator = total_metrics.pop('denominator')
+    return jax.tree.map(lambda x: float(x / eval_denominator), total_metrics)
+
 
   def _eval_batch(self,
```
Comment (rka97) [inline algoperf/workloads/lm/lm_jax/workload.py]: The _eval_batch in the common workload.py should work for here, do we need this function?

File: algoperf/workloads/lm/lm_jax/workload.py (hunk 1 above)
Comment (rka97) [reply algoperf/workloads/lm/lm_jax/workload.py]: Added a commit below to fix this. Maybe we should also use it for WMT? I was getting OOM errors on the V100s without it.

Comment (priyakasimbeg) [reply algoperf/workloads/lm/lm_jax/workload.py]: ok sounds good, thanks. One thing I forgot to mention is that we want to support label smoothing which is another reason I refactored the loss function calls.

File: algoperf/workloads/lm/lm_jax/workload.py (hunk 2 above)
Comment (priyakasimbeg) [reply algoperf/workloads/lm/lm_jax/workload.py]: The weights are used to identify padded elements in the validation split and correctly calculate the number of tokens returned in the eval dict.

File: algoperf/workloads/lm/input_pipeline.py (hunk 3 above)
Comment (priyakasimbeg) [reply algoperf/workloads/lm/input_pipeline.py:69]: No I don't think so. The input_pipeline already calls .repeat() on the train split and we don't want cycle on the validation split.

File: algoperf/workloads/lm/lm_jax/workload.py (hunk 4 above)
Comment (priyakasimbeg) [reply algoperf/workloads/lm/lm_jax/workload.py]: yes you're right, the shared workload version did not return the expected eval_dict. I replaced the shared version with this version now and removed the _eval_batch from the jax-specific workload definition.

File: algoperf/workloads/lm/lm_pytorch/workload.py
Comment (priyakasimbeg) [reply algoperf/workloads/lm/lm_pytorch/workload.py:136]: Left a todo in the pytorch workload to clean this up
